### PR TITLE
Random number generation API redesign

### PR DESCRIPTION
**Pull request by wjakob**
This PR moves the tentative ``dr.rand()`` and ``dr.randn()`` functions into a *generator object* similar to how this works in NumPy and other array APIs. It can be used as follows

```python
rng: dr.random.Generator = dr.rng(seed=123)

value = rng.random(Float, 1000)
value_2 = rng.normal(TensorXf16, shape=(100, 100))
```

Furthermore, the PRNG underlying this interface was switched to a new algorithm named ``Philox4x32``.

This redesign was motivated by the following considerations:

The previously used ``PCG32`` PRNG has very low per-sample cost, but it tends to suffer from correlations between separate parallel streams if not correctly seeded. Another pseudorandom number generator must therefore be added to initialize PCG32 in practice, which increases computational cost and conceptual complexity.

On top of this, the `rng.random()` / `rng.normal()` interface are designed to initialize differently-sized arrays, which means that the PRNG cannot be reused. Therefore, the main benefit of PCG32---low per-sample cost---is not actually that helpful here. This interface is better served by a counter-based PRNG that uses a complex transition function to turn one or more counters into pseudorandom output.

The implementation adopts ``Philox`` from the paper [Parallel Random Numbers: As Easy as 1, 2, 3](https://www.thesalmons.org/john/random123/papers/random123sc11.pdf) by Salmon et al. (2011). This method uses a combination of wide multiplication and XOR operations to transform a seed and four counters into high-quality pseudorandom outputs. Changing either the seed or any of the counters yields a new statistically independent sample. This method is more expensive per sample but relative foolproof, which makes it a better fit for this API. A previous set of changes to Dr.Jit and Dr.Jit-Core added the necessary wide multiplication instructions to compile Philox to efficient machine code.

The commit also adapts functionality using random numbers internally (specifically, ``nn.Module.alloc()``) to optionally take a ``Generator`` object.

### CODE REVIEW

**Review by merlinND — COMMENTED** [review]
Really happy to see this Generator API, thanks a lot for soliciting our feedback!

File: include/drjit/random.h [hunk 1]
```diff
@@ -584,14 +659,262 @@ template <typename T> struct PCG32 {
...
+ *
+ * .. note::
+ *
+ *    For a comparison of use cases between :py:class:`Philox4x32` and
+ *    :py:class:`PCG32`, see the :py:class:`PCG32` class documentation. In
+ *    brief: use :py:class:`PCG32` for sequential generation with lowest cost
+ *    per sample; use :py:class:`Philox4x32` for parallel generation where
+ *    independent streams are critical.
+ */
+template <typename T> struct Philox4x32 {
```
[the remark this answers is not in the document]
Comment (wjakob) [reply include/drjit/random.h:718]: The function test03_philox tests the raw output against OpenRAND.

The strengh-reduced version we use (4x32 with 7 iterations) was verified to pass Test01's BigCrush (https://arxiv.org/abs/1204.6193)

File: drjit/nn.py [hunk 2]
```diff
@@ -87,8 +87,18 @@ def alloc(self, dtype: Type[drjit.ArrayBase], size: int = -1) -> Module:
...
+        Layer weights are initialized using pseudorandom values obtained from
+        the specified generator object ``rng``. If ``rng=None`` (the default),
+        a generator is constructed on the fly. In this case, weight values are
+        consistent across runs (i.e., calling ``alloc()`` twice will produce
+        the same initialization).
         """
-        return self._alloc(dtype, size)[0]
+
+        if rng is None:
+            rng = drjit.rng()
```
[the remark this answers is not in the document]
Comment (wjakob) [reply drjit/nn.py:99]: I don't understand this. I always find these large constants (e.g. for PCG32 seeding) mysterious. It makes people wonder if the quality will suffer if they use a different not-carefully-chosen initialization.

[the remark this answers is not in the document]
Comment (merlinND) [reply drjit/nn.py:99]: I was suggesting to use a "random" / unusual seed for the generator created here (to allocate model weights), in order to avoid this scenario:

```py
my_model = dr.nn.Sequential(...)
my_model = my_model.alloc()   # Not passing generator --> default RNG created with seed 0

# Default-seeded RNG
rng = drjit.rng()

# Inputs to my computation are now 100% correlated with the model's weights :(
result = computation(rng.random(...))
```

The default seed used when calling `dr.rng()` in general can still be 0 / some simple value. I'm only suggesting to use an unusual seed for the default `alloc()` RNG.
